persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,536 @@
+import Replace.Megaparsec (sepCap)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+type Parser = Parsec Void String
+
+-- | Source location: file and line/col information. This is half of a 'Span'.
+data SourceLoc = SourceLoc
+    { locFile :: Text
+    , locStartLine :: Int
+    , locStartCol :: Int
+    }
+    deriving (Show, Lift)
+
+data Token
+    = DocComment Text
+    | Comment Text
+    | Quotation Text
+    | Equality Text Text
+    | Parenthetical Text
+    | BlockKey Text
+    | PText Text
+    deriving (Eq, Ord, Show)
+
+tokenContent :: Token -> Text
+tokenContent = \case
+    Comment s -> s
+    DocComment s -> s
+    Quotation s -> s
+    Equality l r -> mconcat [l, "=", r]
+    Parenthetical s -> s
+    PText s -> s
+    BlockKey s -> s
+
+spaceConsumer :: Parser ()
+spaceConsumer =
+    L.space
+        hspace1
+        (L.skipLineComment "--")
+        empty
+
+spaceConsumerN :: Parser ()
+spaceConsumerN =
+    L.space
+        space1
+        (L.skipLineComment "--")
+        empty
+
+escapedParen :: Parser Char
+escapedParen =
+    char '\\'
+        *> choice
+            [ char '('
+            , char ')'
+            ]
+
+contentChar :: Parser Char
+contentChar =
+    choice
+        [ alphaNumChar
+        , char '.'
+        , char '['
+        , char ']'
+        , char '_'
+        , char '\''
+        , char '"'
+        , char '!'
+        , char '~'
+        , char '-'
+        , char ':'
+        , try escapedParen
+        , char '\\'
+        ]
+
+nonLineSpaceChar :: Parser Char
+nonLineSpaceChar = choice [char ' ', char '\t']
+
+-- This is a replacement for `Text.Megaparsec.Char.Lexer.charLiteral`;
+-- it does nearly the same thing but additionally supports escaped parentheses.
+charLiteral :: Parser Char
+charLiteral = label "literal character" $ do
+    r <- lookAhead (count' 1 10 anySingle)
+    case lexChar r of
+        Just (c, r') -> c <$ skipCount (length r - length r') anySingle
+        Nothing -> unexpected (errorTokens r)
+  where
+    errorTokens ts = Tokens $
+        case NEL.nonEmpty ts of
+            Just nel -> NEL.head nel :| []
+            Nothing -> error "unreachable" -- ts is known to be non-empty
+    lexChar "" =
+        Nothing
+    lexChar ('\\' : '(' : rest) =
+        Just ('(', rest)
+    lexChar ('\\' : ')' : rest) =
+        Just (')', rest)
+    lexChar ('\\' : '\\' : rest) =
+        Just ('\\', rest)
+    lexChar ('\\' : '\'' : rest) =
+        Just ('\'', rest)
+    lexChar ('\\' : '\"' : rest) =
+        Just ('\"', rest)
+    lexChar (a : rest) =
+        Just (a, rest)
+
+equality :: Parser Token
+equality = label "equality expression" $ do
+    L.lexeme spaceConsumer $ do
+        lhs <- some contentChar
+        _ <- char '='
+        rhs <-
+            choice
+                [ try quotation'
+                , try sqlLiteral
+                , try parentheticalInner
```

Comment:
Is `try` necessary inside of `choice`? that feels vaguely like an antipattern - there's a blog post (and [reddit discussion](https://www.reddit.com/r/haskell/comments/25ulsv/ezyang_try_a_b_considered_harmful/)) about this pattern which suggests ways to potentially improve this

Reply:
I think these `try`s may simply be unnecessary. If that turns out to be true, I'll just delete them.